keps/sig-api-machinery/1904-efficient-watch-resumption/README.md

```diff
@@ -0,0 +1,485 @@
+- on start, quorum list is send to etcd
+- the watch cache is set to be synced to resource version of that list
+- from that point, watch to etcd is established and watch cache is
+  updated by incoming watch event stream; note that only objects of
+  a given resource type are watched
+- if watch cannot be reestablished, all watchers connected to it are
+  disconnected, a new quorum list is send and watchcache is reset to
+  that point in time
+
+It's important to note that while only objects of a given resource type
+are watched (thus resource version to which watch cache is synced can
+only be updated to a value that some object of that type had), the quorum
+list, even though it is only requesting objects of that resource type,
+returns "globally current version for the whole etcd".
+
+While the inability to reestablish watch doesn't happen often in practice,
+the kube-apiserver restarts obviously does (e.g. on upgrades). And this
+is resulting in two main problems:
+
+1. In case of rolling upgrade of kube-apiserver, when no object of a given
+   resource type is changing (but objects of other types do), all watchers
+   will eventually be forced to relist, causing significant performance
+   and scalability issues for larger clusters. See the example below for
+   clarification.
+
+1. If no object of a given resource type is changing after kube-apiserver
+  startup, watch cache may stuck being synced to resource version not being
+	version of any object of a given type for extended period of time. See
+  https://github.com/kubernetes/kubernetes/issues/91073 for more details.
+
+To illustrate the problem better, imagine the following example:
+- there is a single kube-apiserver that we are going to restart
+- there are only two resource types: X and Y (for simplicity)
+- there is one object of type X: x with rv=100
+- there is one object of type Y: y with rv=101
+- there is a watch W for objects of type X already synced to rv=100
+- watchcache for both X and Y is up-to-date
+
+- T1: kube-apiserver being restarted
+- T2: kube-apiserver is initialized with watchcache for X synced to
+   rv=101 (most recent etcd version)
+- T3: watch W is trying to reconnect, given rv=100 is outside of
+   supported history window (we only support versions from 101 being
+   the moment of watch cache initialization), relist is forced
+
+Note that adding more kube-apiservers doesn't solve the problem. In
+fact, it is introducing the second problem where watch cache in different
+instances may be out of sync for long period of time. To achieve such
+situation it is enough to have object y be updated to rv=102 between
+the moment kube-apiserver-1 is initialized and kube-apiserver-2 starts
+its initialization. And we would stuck with kube-apiserver-1 synced to
+rv=101 and kube-apiserve-2 synced to rv=102 until the next change of
+any object of type X happens (even though both watch caches contain the
```

Comment:
OK, you're talking about something I guess I didn't understand from the first draft, and actually I still don't think I understand it.

Are you saying that a low-churn resource type could get stuck in the distant past due to an unrelated high-churn type?

Whatever this second problem is, does it actually cause different effects than the first problem?

Reply:
It's not exactly like that.

I have updated this part of clear example (without mixing it with the first problem).
Is it more clear now?